crates/sdk/src/masp.rs

```diff
@@ -753,33 +881,54 @@ impl<U: ShieldedUtils + MaybeSend + MaybeSync> ShieldedContext<U> {
         // persist the cache in case of interruptions.
         let _ = self.save().await;
 
-        let txs = logger.scan(self.unscanned.clone());
-        for (indexed_tx, (epoch, tx, stx)) in txs {
-            if Some(indexed_tx) > last_witnessed_tx {
-                self.update_witness_map(indexed_tx, &stx)?;
-            }
-            let mut vk_heights = BTreeMap::new();
-            std::mem::swap(&mut vk_heights, &mut self.vk_heights);
-            for (vk, h) in vk_heights
-                .iter_mut()
-                .filter(|(_vk, h)| **h < Some(indexed_tx))
-            {
-                self.scan_tx(
-                    indexed_tx,
-                    epoch,
-                    &tx,
-                    &stx,
-                    vk,
-                    native_token.clone(),
-                )?;
-                *h = Some(indexed_tx);
+        let unscanned = self.unscanned.clone();
+        for (indexed_tx, (_, _, stx)) in &unscanned.txs {
+            if Some(*indexed_tx) > last_witnessed_tx {
+                self.update_witness_map(*indexed_tx, stx)?;
             }
-            // possibly remove unneeded elements from the cache.
-            self.unscanned.scanned(&indexed_tx);
-            std::mem::swap(&mut vk_heights, &mut self.vk_heights);
-            let _ = self.save().await;
         }
+        _ = self.utils.save(self).await;
+
+        let sync_status = self.sync_status;
+        let txs = logger.scan(unscanned);
+        let vk_heights = self.vk_heights.clone();
+        let ctx = ThreadHandle::new(self);
+        txs.par_bridge()
```

Comment:
Is it possible to interrupt this process? Or the interrupt between the `update_witness_map` and the call to rayon here? Cause it seems to me like we could call `merge_scanned_data` out of order and maybe this could be a problem if the process gets interrupted cause we might have an inconsistent state of the shielded context?